File: ConstructBinaryTreefromInorderandPostorderTraversal2.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None
        
class Solution:
    def buildTree(self, inorder, postorder):
        """
        :type preorder: List[int]
        :type inorder: List[int]
        :rtype: TreeNode
        """
        if len(inorder)==0:
            return None
            
        i = len(postorder)-1
        j = len(inorder)-1
        
        isLeft = False
        
        stack = []
        root = TreeNode(postorder[i])
        t = root
        stack.append(t)
        i -= 1
        while i>=0:
            if len(stack)>0 and stack[-1].val==inorder[j]:
                t = stack.pop()
                isLeft = True
                j -= 1
            else:
                if isLeft:
                    t.left = TreeNode(postorder[i])
                    t = t.left
                    stack.append(t)
                    isLeft = False
                    i -= 1
                else:
                    t.right = TreeNode(postorder[i])
                    t = t.right
                    stack.append(t)
                    i -= 1
        return root
        
    def printTree(self, root):
        if root!=None:
            print(root.val)
            self.printTree(root.left)
            self.printTree(root.right)
        
if __name__=='__main__':
    inorder = list(input())
    postorder = list(input())
    
    solution = Solution()
    root = solution.buildTree(inorder, postorder)
    solution.printTree(root)

File: ConstructBinaryTreefromInorderandPostorderTraversal2.cpp
#include<iostream>
#include<vector>

using namespace std;

struct TreeNode{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        if(inorder.empty()) return NULL;
        
        int i=postorder.size()-1, j=inorder.size()-1;
        vector<TreeNode*> stack;
        
        TreeNode *root, *t;
        
        bool isLeft=false;
        
        root=new TreeNode(postorder[i]);
        t=root;
        stack.push_back(root);
        i--;
        
        while(i>=0){
            
            if(!stack.empty()&&stack.back()->val==inorder[j]){
                t=stack.back();
                stack.pop_back();
                j--;
                isLeft=true;
            }else{
                if(isLeft){
                    t->left=new TreeNode(postorder[i]);
                    t=t->left;
                    stack.push_back(t);
                    isLeft=false;
                    i--;
                }else{
                    t->right=new TreeNode(postorder[i]);
                    t=t->right;
                    stack.push_back(t);
                    i--;
                }
            }
        }
        
        return root;
    }
    
    void printTree(TreeNode *root){
        if(root!=NULL){
            cout<<root->val<<" ";
            printTree(root->left);
            printTree(root->right);
            cout<<endl;
        }
    }
};

int main(){
    
    int n;
    
    cin>>n;
    
    vector<int> inorder;
    vector<int> postorder;
    
    for(int i=0;i<n;i++){
        int num;
        cin>>num;
        inorder.push_back(num);
    }
    
    for(int i=0;i<n;i++){
        int num;
        cin>>num;
        postorder.push_back(num);
    }
    
    Solution *solution=new Solution();
    TreeNode *root=solution->buildTree(inorder, postorder);
    
    solution->printTree(root);
    
    return 0;
}
